examples/storm-starter/src/jvm/org/apache/storm/starter/PersistentWindowingTopology.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.storm.streams.Pair;
+import org.apache.storm.task.OutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.topology.OutputFieldsDeclarer;
+import org.apache.storm.topology.TopologyBuilder;
+import org.apache.storm.topology.base.BaseStatefulWindowedBolt;
+import org.apache.storm.tuple.Fields;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.windowing.TupleWindow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An example that demonstrates the usage of {@link org.apache.storm.topology.IStatefulWindowedBolt} with window
+ * persistence.
+ * <p>
+ * The framework automatically checkpoints the tuples in window along with the bolt's state and restores the same
+ * during restarts.
+ * </p>
+ *
+ * <p>
+ * This topology uses 'redis' for state persistence, so you should also start a redis instance before deploying.
+ * If you are running in local mode you can just start a redis server locally which will be used for storing the state. The default
+ * RedisKeyValueStateProvider parameters can be overridden by setting "topology.state.provider.config", for e.g.
+ * <pre>
+ * {
+ *   "jedisPoolConfig": {
+ *     "host": "redis-server-host",
+ *     "port": 6379,
+ *     "timeout": 2000,
+ *     "database": 0,
+ *     "password": "xyz"
+ *   }
+ * }
+ * </pre>
+ * </p>
+ */
+public class PersistentWindowingTopology {
+    private static final Logger LOG = LoggerFactory.getLogger(PersistentWindowingTopology.class);
+
+    // wrapper to hold global and window averages
+    private static class Averages {
+        private final double global;
+        private final double window;
+
+        Averages(double global, double window) {
+            this.global = global;
+            this.window = window;
+        }
+
+        @Override
+        public String toString() {
+            return "Averages{" + "global=" + String.format("%.2f", global) + ", window=" + String.format("%.2f", window) + '}';
+        }
+    }
+
+    /**
+     * A bolt that uses stateful persistence to store the windows along with the state (global avg).
+     */
+    private static class AvgBolt extends BaseStatefulWindowedBolt<KeyValueState<String, Pair<Long, Long>>> {
+        private static final String STATE_KEY = "avg";
+
+        private OutputCollector collector;
+        private KeyValueState<String, Pair<Long, Long>> state;
+        private Pair<Long, Long> globalAvg;
+
+        @Override
+        public void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector collector) {
+            this.collector = collector;
+        }
+
+        @Override
+        public void initState(KeyValueState<String, Pair<Long, Long>> state) {
+            this.state = state;
+            globalAvg = state.get(STATE_KEY, Pair.of(0L, 0L));
+            LOG.info("initState with global avg [" + (double) globalAvg._1 / globalAvg._2 + "]");
```

Comment:
_1 and _2 are exposed to be consistent with other `TupleN` classes. Will use the getters here.